corehq/sql_db/jsonops.py

```diff
@@ -0,0 +1,69 @@
+import json
+
+from django.db.models.expressions import Expression, Func, RawSQL, Value
+from django.db.models import JSONField
+
+
+class JsonDelete(Func):
+    """Delete item(s) from JSONB value
+
+    :param expression: JSONB field name or value expression.
+    :param *items: Names of items to delete.
+    """
+    function = "-"
+    template = "%(expressions)s::jsonb %(function)s '{%(items)s}'::text[]"
+    arity = 1
+
+    def __init__(self, expression, *items):
+        items = ",".join(items)
+        if "'" in items:
+            raise ValueError(f"invalid items: {items!r}")
+        super().__init__(expression, output_field=JSONField(), items=items)
+
+
+class JsonGet(Func):
+    """Get item from JSON or JSONB value
+
+    :param expression: JSON/JSONB field name or value expression.
+    :param field: Name of item to get.
+    """
+    function = "->"
+    template = "%(expressions)s%(function)s'%(field_name)s'"
+    arity = 1
+
+    def __init__(self, expression, field):
+        if not isinstance(field, str) or "'" in field:
+            raise ValueError(f"invalid field: {field!r}")
+        super().__init__(expression, output_field=JSONField(), field_name=field)
+
+
+class JsonSet(Func):
+    """Replace value at path in JSONB value
+
+    :param expression: JSONB field name or value expression.
+    :param path: Sequence of keys or indexes representing the path of
+        the value in `expression` to be replaced.
+    :param new_value: JSON-serializable value or expression.
+    :param create_missing: Add value if it is missing (default: True).
+    """
```

Comment:
a link to the PG docs or else an example of what this renders to would be useful.

Reply:
59b3dcd2fb561014a8b7c8625353d0ffb3922ee1